### DISCUSSION TITLE
Optimising compile times

### DISCUSSION THREAD

**Discussion by arlyon** [arlyon/async-stripe#77] [General]
It's no secret that this library has quite terrible compile perf:

- high compile times
- very large binaries
- high degree of monomorphisation

Here is some data, based on the benchmark project (which still uses the blocking runtime):

```bash
export RUNTIME=runtime-blocking

# total code size for release builds
> cargo bloat --features $RUNTIME -n 0 --release --message-format json | jq '.["text-section-size"]'
4283321 (≈4mb)

# size of stripe library
> cargo bloat --features $RUNTIME --crates -n 0 --release --message-format json | jq '.crates[] | select(.name == "stripe") | .size'
3036152 (≈2.9mb)

# size of serde code in stripe
> cargo bloat --features $RUNTIME -n 0 --release --message-format json | jq '.functions | map(select(.crate == "stripe")) | map(select(.name | contains("serde"))) |  map(.size) | add'
2157228 (≈2.0mb)

# size of stripe library excluding dependencies
> cargo bloat --features $RUNTIME -n 0 --release --message-format json | jq '[ .functions[] | select(.crate == "stripe") | select(.name | contains("stripe")) | select(.name | contains("serde") | not) | .size ] | add'
67602 (≈60kb)

# size of all crates
> cargo bloat --features $RUNTIME -n 0 --crates --release --message-format json | jq -r '.crates[] | "\(.name): \(.size)"'
stripe: 3036152
std: 600365
h2: 222939
hyper: 114707
tokio: 92038
http: 45162
serde_json: 19001
serde: 15145
hashbrown: 13672
openssl: 11746
bytes: 5794
mio: 4170
tracing_core: 3788
httparse: 2992
serde_qs: 2730
openssl_probe: 2704
tokio_util: 2446
native_tls: 2428
futures_channel: 2160
binary_size: 1976
indexmap: 1784
tracing: 1567
socket2: 1233
hyper_tls: 1212
want: 1117
smol_str: 471
openssl_sys: 338
futures_util: 336
percent_encoding: 319
[Unknown]: 298
futures_core: 276
log: 179
futures_task: 21
tokio_native_tls: 8
```

Clearly, from this, serde is vastly increasing the size of the code, and is likely the primary culprit. Solutions?

- [x] aggressive feature gating
- [ ] reduce codegen and monomorphisation (miniserde?)
  - miniserde does not support enums, only structs

**Comment by arlyon** [comment]
I will say as well, looking at the output of `cargo expand`, we can see the total cost of the auto-generated code:

```
> tokei src
38209
> cargo expand | wc -l
635395
```

Obviously we can expect the runtime performance of `cargo check` to scale linearly with loc in the best case and this library is, once all the macros are expanded, 630k lines of code. Huge! For the question of compile times we can look at splitting out sub-crates, but the loc count isn't going to improve by much as long as we are using serde.

**Reply by arlyon** [reply]
As of 27/01/22 the code is now 730432 loc

**Comment by arlyon** [comment]
Had some more digging today, this time using `-Z time-passes`:

```
time:   9.472; rss:   51MB ->  598MB ( +547MB)	expand_crate
time:   0.000; rss:  598MB ->  598MB (   +0MB)	check_unused_macros
time:   9.472; rss:   51MB ->  598MB ( +547MB)	macro_expand_crate
time:   0.000; rss:  598MB ->  598MB (   +0MB)	maybe_building_test_harness
time:   0.067; rss:  598MB ->  595MB (   -4MB)	AST_validation
time:   0.000; rss:  595MB ->  595MB (   +0MB)	maybe_create_a_macro_crate
time:   0.004; rss:  595MB ->  595MB (   +0MB)	finalize_imports
time:   0.023; rss:  595MB ->  599MB (   +4MB)	finalize_macro_resolutions
time:   0.764; rss:  599MB ->  822MB ( +223MB)	late_resolve_crate
time:   0.000; rss:  822MB ->  822MB (   +0MB)	resolve_main
time:   0.044; rss:  822MB ->  822MB (   +0MB)	resolve_check_unused
time:   0.000; rss:  822MB ->  822MB (   +0MB)	resolve_report_errors
time:   0.041; rss:  822MB ->  822MB (   +0MB)	resolve_postprocess
time:   0.876; rss:  595MB ->  822MB ( +227MB)	resolve_crate
time:   0.045; rss:  822MB ->  822MB (   +0MB)	complete_gated_feature_checking
time:  10.460; rss:   51MB ->  822MB ( +771MB)	configure_and_expand
time:   0.000; rss:  822MB ->  822MB (   +0MB)	prepare_outputs
time:   0.000; rss:  822MB ->  822MB (   +0MB)	blocked_on_dep_graph_loading
time:   1.212; rss:  822MB -> 1162MB ( +341MB)	hir_lowering
time:   0.239; rss: 1162MB -> 1152MB (  -10MB)	early_lint_checks
time:   0.125; rss: 1152MB -> 1152MB (   +0MB)	drop_ast
time:   0.000; rss: 1065MB -> 1065MB (   +0MB)	setup_global_ctxt
time:   0.000; rss: 1069MB -> 1069MB (   +0MB)	looking_for_entry_point
time:   0.013; rss: 1069MB -> 1069MB (   +0MB)	looking_for_derive_registrar
time:   0.892; rss: 1069MB -> 1096MB (  +27MB)	misc_checking_1
time:   0.789; rss: 1096MB -> 1123MB (  +26MB)	type_collecting
time:   0.033; rss: 1123MB -> 1123MB (   +0MB)	impl_wf_inference
time:   0.003; rss: 1173MB -> 1173MB (   +0MB)	unsafety_checking
time:   0.006; rss: 1173MB -> 1173MB (   +0MB)	orphan_checking
time:   0.144; rss: 1123MB -> 1173MB (  +50MB)	coherence_checking
time:   2.104; rss: 1173MB -> 1224MB (  +51MB)	wf_checking
time:   0.638; rss: 1224MB -> 1186MB (  -39MB)	item_types_checking
time:  12.394; rss: 1186MB -> 1140MB (  -45MB)	item_bodies_checking
time:  16.134; rss: 1096MB -> 1142MB (  +46MB)	type_check_crate
time:   0.391; rss: 1142MB -> 1145MB (   +3MB)	match_checking
time:   0.299; rss: 1145MB -> 1151MB (   +5MB)	liveness_and_intrinsic_checking
time:   0.691; rss: 1142MB -> 1151MB (   +9MB)	misc_checking_2
time:  11.072; rss: 1151MB -> 1551MB ( +401MB)	MIR_borrow_checking
time:   0.051; rss: 1551MB -> 1552MB (   +1MB)	MIR_effect_checking
time:   0.000; rss: 1552MB -> 1552MB (   +0MB)	layout_testing
time:   0.231; rss: 1554MB -> 1612MB (  +58MB)	death_checking
time:   0.001; rss: 1612MB -> 1612MB (   +0MB)	unused_lib_feature_checking
time:   0.482; rss: 1612MB -> 1637MB (  +25MB)	crate_lints
time:   0.327; rss: 1637MB -> 1637MB (   +0MB)	module_lints
time:   0.809; rss: 1612MB -> 1637MB (  +25MB)	lint_checking
time:   0.547; rss: 1637MB -> 1637MB (   +0MB)	privacy_checking_modules
time:   2.042; rss: 1552MB -> 1637MB (  +85MB)	misc_checking_3
time:   0.023; rss: 1813MB -> 1813MB (   +0MB)	monomorphization_collector_root_collections
time:   4.074; rss: 1813MB -> 1940MB ( +126MB)	monomorphization_collector_graph_walk
time:   0.966; rss: 1940MB -> 1972MB (  +32MB)	partition_and_assert_distinct_symbols
time:  10.330; rss: 1637MB -> 1969MB ( +332MB)	generate_crate_metadata
time:   0.000; rss: 1963MB -> 1963MB (   +0MB)	find_cgu_reuse
time:  24.867; rss: 1963MB -> 2229MB ( +266MB)	codegen_to_LLVM_IR
time:  24.976; rss: 1969MB -> 2229MB ( +259MB)	codegen_crate
time:   0.000; rss: 2229MB -> 2229MB (   +0MB)	assert_dep_graph
time:   0.000; rss: 2229MB -> 2229MB (   +0MB)	check_dirty_clean
time:   0.016; rss: 2229MB -> 2228MB (   -1MB)	encode_query_results_for(rustc_query_impl::queries::type_of)
time:   0.024; rss: 2228MB -> 2229MB (   +1MB)	encode_query_results_for(rustc_query_impl::queries::generics_of)
time:  23.106; rss: 2091MB -> 2232MB ( +141MB)	LLVM_passes(crate)
time:   0.011; rss: 2229MB -> 2232MB (   +3MB)	encode_query_results_for(rustc_query_impl::queries::predicates_of)
time:   0.001; rss: 2232MB -> 2225MB (   -7MB)	encode_query_results_for(rustc_query_impl::queries::mir_const_qualif)
time:   0.007; rss: 2225MB -> 2225MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::mir_for_ctfe)
time:   0.362; rss: 2225MB -> 2225MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::optimized_mir)
time:   0.000; rss: 2225MB -> 2225MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::covered_file_name)
time:   0.000; rss: 2225MB -> 2225MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::covered_code_regions)
time:   0.012; rss: 2225MB -> 2225MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::promoted_mir)
time:   0.001; rss: 2225MB -> 2225MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::unsafety_check_result)
time:   0.000; rss: 2225MB -> 2225MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::thir_check_unsafety)
time:   0.114; rss: 2225MB -> 2225MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::typeck)
time:   0.000; rss: 2225MB -> 2225MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::diagnostic_only_typeck)
time:   0.001; rss: 2225MB -> 2225MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::used_trait_imports)
time:   0.001; rss: 2225MB -> 2225MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::mir_borrowck)
time:   0.001; rss: 2225MB -> 2225MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::eval_to_allocation_raw)
time:   0.003; rss: 2225MB -> 2225MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::eval_to_const_value_raw)
time:   0.000; rss: 2225MB -> 2225MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::check_match)
time:   0.011; rss: 2225MB -> 2225MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::symbol_name)
time:   0.007; rss: 2225MB -> 2225MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::codegen_fn_attrs)
time:   0.012; rss: 2225MB -> 2229MB (   +4MB)	encode_query_results_for(rustc_query_impl::queries::codegen_fulfill_obligation)
time:   0.005; rss: 2229MB -> 2229MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::specialization_graph_of)
time:   0.000; rss: 2229MB -> 2230MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::adt_drop_tys)
time:   0.000; rss: 2230MB -> 2230MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::adt_significant_drop_tys)
time:   0.000; rss: 2230MB -> 2230MB (   +0MB)	encode_query_results_for(rustc_query_impl::queries::unused_generic_params)
time:   0.591; rss: 2229MB -> 2230MB (   +1MB)	encode_query_results
time:   0.669; rss: 2229MB -> 2225MB (   -3MB)	incr_comp_serialize_result_cache
time:   0.669; rss: 2229MB -> 2225MB (   -3MB)	incr_comp_persist_result_cache
time:   0.000; rss: 2225MB -> 2225MB (   +0MB)	incr_comp_persist_dep_graph
time:   0.670; rss: 2229MB -> 2225MB (   -3MB)	serialize_dep_graph
time:   0.204; rss: 2225MB -> 1736MB ( -489MB)	free_global_ctxt
time:   0.000; rss: 1736MB -> 1736MB (   +0MB)	join_worker_thread
time:   0.206; rss: 1736MB -> 1736MB (   +0MB)	copy_all_cgu_workproducts_to_incr_comp_cache_dir
time:   0.206; rss: 1736MB -> 1736MB (   +0MB)	finish_ongoing_codegen
time:   0.000; rss: 1736MB -> 1736MB (   +0MB)	llvm_dump_timing_file
time:   0.003; rss: 1736MB -> 1700MB (  -36MB)	serialize_work_products
time:   0.000; rss: 1595MB -> 1595MB (   +0MB)	incr_comp_finalize_session_directory
time:   0.001; rss: 1595MB -> 1595MB (   +0MB)	link_binary_check_files_are_writeable
time:   0.438; rss: 1595MB -> 1580MB (  -16MB)	link_rlib
time:   0.004; rss: 1580MB -> 1580MB (   +0MB)	link_binary_remove_temps
time:   0.453; rss: 1595MB -> 1580MB (  -16MB)	link_binary
time:   0.459; rss: 1595MB -> 1438MB ( -158MB)	link_crate
time:   0.675; rss: 1736MB -> 1438MB ( -299MB)	link
time:  80.185; rss:   40MB ->  165MB ( +125MB)	total
```

Sorting by time taken:

```bash
> sort -nk 2 times.txt | tail -11 # top 10 plus total
time:   9.472; rss:   51MB ->  598MB ( +547MB)  expand_crate
time:   9.472; rss:   51MB ->  598MB ( +547MB)  macro_expand_crate
time:  10.330; rss: 1637MB -> 1969MB ( +332MB)  generate_crate_metadata
time:  10.460; rss:   51MB ->  822MB ( +771MB)  configure_and_expand
time:  11.072; rss: 1151MB -> 1551MB ( +401MB)  MIR_borrow_checking
time:  12.394; rss: 1186MB -> 1140MB (  -45MB)  item_bodies_checking
time:  16.134; rss: 1096MB -> 1142MB (  +46MB)  type_check_crate
time:  23.106; rss: 2091MB -> 2232MB ( +141MB)  LLVM_passes(crate)
time:  24.867; rss: 1963MB -> 2229MB ( +266MB)  codegen_to_LLVM_IR
time:  24.976; rss: 1969MB -> 2229MB ( +259MB)  codegen_crate
time:  80.185; rss:   40MB ->  165MB ( +125MB)	total
```

Interestingly, we peak at over 2GB of RAM when compiling as well.

**Comment by FL33TW00D** [comment]
Just to add some more to the above:
```rust
#[derive(serde::Deserialize, serde::Serialize)]
pub struct Discount {
    // Always true for a deleted object
    #[serde(default)]
    pub deleted: bool,
    /// The subscription that this coupon is applied to, if it is applied to a particular subscription.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subscription: Option<String>,
}

fn main() {}
```
`serde-1.0.79`
`cargo expand | wc -l` produces `304` lines.

Switching over to `miniserde` reduces to `104` lines, but it doesn't support the `default` or `skip_serializing_if` used in the toy example.

Based on the above findings and this, we can see that serde is producing almost all of the bloat in the crate. One solution we've discussed is simply splitting the crate into multiple crates based on the existing feature gates, eg (`async-stripe-billing`?)

**Comment by FL33TW00D** [comment]
In light of [this](https://github.com/serde-rs/serde/pull/2250) recent PR on serde, I thought it would be interesting to compare with the old results.

The below was done with `serde 1.0.136`:
`tokei src`-> `60866` lines.
`cargo expand --features runtime-blocking | wc -l` -> `1025473` lines total.

Therefore we are seeing near enough 1 million lines of serde generated code.

The below was done with a patch of (serde)[https://github.com/nnethercote/serde/commit/f8cc37c8547589edb88123424b55cce19a5feaaf] from the PR:
`cargo expand --features runtime-blocking | wc -l` -> `974076` lines total.

Delta of `51397` lines, or a `5%` reduction.

**Reply by arlyon** [reply]
nice! thanks. We should get this 'for free' since we only have a minimum bound on serde

**Comment by AThilenius** [comment]
Please don't take this as criticism, just as a data point. We are pulling `async-stripe` from our deps and replacing it with manual `reqwest` calls because of build times. My 0.02 is that this should be highly prioritized, but I might be alone in that opinion 🤷‍♂️

**Reply by arlyon** [reply]
That's understandable. I am not aware of any large 'api crates' that use serde and don't suffer from massive compile times and if your api surface is small enough it may make sense not to take that hit.

We have experimented a little with miniserde which features significantly less monomorphisation, but without support for tagged enums there isn't a viable out-of-the-box solution to get around it.

Stripe's API is very deeply intertwined so splitting into crates in an automated way is nontrivial.

**Reply by arlyon** [reply]
That said, I have had a project in the back of my mind to support manually spitting out impls for Deserialize for those parts of the api that aren't derivable for a while now but haven't had a strong enough urge to do this since for my personal use-case what we have now is 'good enough'. It is however on the roadmap :) Thanks for chiming in.

**Reply by AThilenius** [reply]
Our use case is very limited, just a handful of calls. I think a fair bit of "blame" (if that's the right word) can be places on the Rust toolchain here too. I love Rust, but I frequently lament the slow compile times. Switching to Go/C# briefly is enough to remind me just how bad they are lol. That and the decisions made around floating point comparisons. Both I can live with though, so 🤷‍♂️

**Reply by Redrield** [reply]
I'm in the same boat, the compilation requirements of this library breaks the box running our github actions CD, so I ended up just tearing it out and replacing it with reqwest calls to the few endpoints we actually use

**Reply by arlyon** [reply]
We are working on a PR that replaces serde with miniserde here #314, so I would hope this would be mostly 'fixed' once that lands.

**Reply by iganev** [reply]
Broke several ec2 instances with GitLab runners in one day. The machine runs out of RAM and the server explodes. I tried to trim down the features and disable default features - no luck. Even some teammate computers can't handle compiling the library without freezing.

**Comment by WesleyAC** [comment]
As a user of this crate, should I expect turning off unused features to significantly reduce compile times?

**Comment by arlyon** [comment]
Hey folks, some lovely numbers from the rewrite (currently hiding in the `next` branch).

>For some quite unscientific timings, a clean release build of examples/endpoints goes from ~4m to 50s with min-ser enabled. More importantly though, the actual time to build just the binary goes from 75s to 7s, so incremental builds for code depending on async-stripe should be much faster.

> Stripped binary size of the examples/endpoints binary also went from ~70MB to ~20MB. There's likely room for further binary size improvement since a fat LTO build shrunk this further to ~13 MB.

It is not ready for production but we are getting close to an RC

**Reply by WesleyAC** [reply]
@arlyon This is super exciting news, thanks for the work on this! Any idea when a release candidate is expected?

**Comment by adamchalmers** [comment]
Wow, that's awesome Alexander. Thank you so much for the work on this!

**Comment by pikitgb** [comment]
Not an expert on rust at all, but searching around came across this discussion.

Compiling this on a t2.small doesn't work, it doesn't work also in t2.media (about 4 GB ) or ram. It just eats the RAM pretty quick until the computer becomes unresponsive, you can see with htop in ubuntu that it builds up in a few seconds and reaches the 4 gigs

You have to stop the EC2 instance manually from AWS, and run again. 

It sounds that it might be just better for now not to use this crate ? and use an HTTP client and talk to Stripe API directly for the use-cases of each app.

Not sure if this is helpful:

```
/home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustc --crate-name s
tripe --edition=2021 /home/ubuntu/.cargo/registry/src/index.crates.io-6f17d22bba15001f/async-stripe-0.31.2/src/lib.rs
 --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=109 --crate-type li
b --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg feature="async" --cfg feature="billing" --cf
g feature="checkout" --cfg feature="chrono" --cfg feature="connect" --cfg feature="default" --cfg feature="events" --
cfg feature="fraud" --cfg feature="full" --cfg feature="futures-util" --cfg feature="hex" --cfg feature="hmac" --cfg
feature="hyper" --cfg feature="hyper-client" --cfg feature="hyper-tls" --cfg feature="issuing" --cfg feature="orders"
 --cfg feature="runtime-tokio-hyper" --cfg feature="sha2" --cfg feature="sigma" --cfg feature="stream" --cfg feature=
"tax-calculation" --cfg feature="terminal" --cfg feature="tokio" --cfg feature="uuid" --cfg feature="webhook-endpoint
s" --cfg feature="webhook-events" -C metadata=4104e96c465a94fc -C extra-filename=-4104e96c465a94fc --out-dir /home/ub
untu/all_websites/target/debug/deps -C linker=x86_64-linux-gnu-gcc -L dependency=/home/ubuntu/all_websites/target/deb
ug/deps --extern chrono=/home/ubuntu/all_websites/target/debug/deps/libchrono-e6f76212bb01f92c.rmeta --extern futures
_util=/home/ubuntu/all_websites/target/debug/deps/libfutures_util-eab0ff9fe4613bc2.rmeta --extern hex=/home/ubuntu/al
l_websites/target/debug/deps/libhex-bcb0a1ba279f6956.rmeta --extern hmac=/home/ubuntu/all_websites/target/debug/deps/
libhmac-fbc735e0a96a1c80.rmeta --extern http_types=/home/ubuntu/all_websites/target/debug/deps/libhttp_types-d7ee33b5
80be457c.rmeta --extern hyper=/home/ubuntu/all_websites/target/debug/deps/libhyper-eded9a0aa54f854c.rmeta --extern hy
per_tls=/home/ubuntu/all_websites/target/debug/deps/libhyper_tls-fe2033d8f4ffb278.rmeta --extern serde=/home/ubuntu/a
ll_websites/target/debug/deps/libserde-4a0cddb34202a536.rmeta --extern serde_json=/home/ubuntu/all_websites/target/de
bug/deps/libserde_json-288673135d6dc5ad.rmeta --extern serde_path_to_error=/home/ubuntu/all_websites/target/debug/dep
s/libserde_path_to_error-06d247cbee56e73f.rmeta --extern serde_qs=/home/ubuntu/all_websites/target/debug/deps/libserd
e_qs-034782d3eb097a01.rmeta --extern sha2=/home/ubuntu/all_websites/target/debug/deps/libsha2-385537d261f23a69.rmeta
--extern smart_default=/home/ubuntu/all_websites/target/debug/deps/libsmart_default-13b64a7cd8d27eb5.so --extern smol
_str=/home/ubuntu/all_websites/target/debug/deps/libsmol_str-f10535a685011dcf.rmeta --extern thiserror=/home/ubuntu/a
ll_websites/target/debug/deps/libthiserror-7a1530efc030e51c.rmeta --extern tokio=/home/ubuntu/all_websites/target/deb
ug/deps/libtokio-d9d9392c63bfd1e7.rmeta --extern uuid=/home/ubuntu/all_websites/target/debug/deps/libuuid-f2622212b77
0dcf3.rmeta --cap-lints allow -C link-arg=-fuse-ld=lld
```

**Reply by iganev** [reply]
> Not an expert on rust at all, but searching around came across this discussion.
> 
> Compiling this on a t2.small doesn't work, it doesn't work also in t2.media (about 4 GB ) or ram. It just eats the RAM pretty quick until the computer becomes unresponsive, you can see with htop in ubuntu that it builds up in a few seconds and reaches the 4 gigs
> 
> You have to stop the EC2 instance manually from AWS, and run again.
> 
> It sounds that it might be just better for now not to use this crate ? and use an HTTP client and talk to Stripe API directly for the use-cases of each app.
> 
> Not sure if this is helpful:
> 
> ```
> /home/ubuntu/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustc --crate-name s
> tripe --edition=2021 /home/ubuntu/.cargo/registry/src/index.crates.io-6f17d22bba15001f/async-stripe-0.31.2/src/lib.rs
>  --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=109 --crate-type li
> b --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg feature="async" --cfg feature="billing" --cf
> g feature="checkout" --cfg feature="chrono" --cfg feature="connect" --cfg feature="default" --cfg feature="events" --
> cfg feature="fraud" --cfg feature="full" --cfg feature="futures-util" --cfg feature="hex" --cfg feature="hmac" --cfg
> feature="hyper" --cfg feature="hyper-client" --cfg feature="hyper-tls" --cfg feature="issuing" --cfg feature="orders"
>  --cfg feature="runtime-tokio-hyper" --cfg feature="sha2" --cfg feature="sigma" --cfg feature="stream" --cfg feature=
> "tax-calculation" --cfg feature="terminal" --cfg feature="tokio" --cfg feature="uuid" --cfg feature="webhook-endpoint
> s" --cfg feature="webhook-events" -C metadata=4104e96c465a94fc -C extra-filename=-4104e96c465a94fc --out-dir /home/ub
> untu/all_websites/target/debug/deps -C linker=x86_64-linux-gnu-gcc -L dependency=/home/ubuntu/all_websites/target/deb
> ug/deps --extern chrono=/home/ubuntu/all_websites/target/debug/deps/libchrono-e6f76212bb01f92c.rmeta --extern futures
> _util=/home/ubuntu/all_websites/target/debug/deps/libfutures_util-eab0ff9fe4613bc2.rmeta --extern hex=/home/ubuntu/al
> l_websites/target/debug/deps/libhex-bcb0a1ba279f6956.rmeta --extern hmac=/home/ubuntu/all_websites/target/debug/deps/
> libhmac-fbc735e0a96a1c80.rmeta --extern http_types=/home/ubuntu/all_websites/target/debug/deps/libhttp_types-d7ee33b5
> 80be457c.rmeta --extern hyper=/home/ubuntu/all_websites/target/debug/deps/libhyper-eded9a0aa54f854c.rmeta --extern hy
> per_tls=/home/ubuntu/all_websites/target/debug/deps/libhyper_tls-fe2033d8f4ffb278.rmeta --extern serde=/home/ubuntu/a
> ll_websites/target/debug/deps/libserde-4a0cddb34202a536.rmeta --extern serde_json=/home/ubuntu/all_websites/target/de
> bug/deps/libserde_json-288673135d6dc5ad.rmeta --extern serde_path_to_error=/home/ubuntu/all_websites/target/debug/dep
> s/libserde_path_to_error-06d247cbee56e73f.rmeta --extern serde_qs=/home/ubuntu/all_websites/target/debug/deps/libserd
> e_qs-034782d3eb097a01.rmeta --extern sha2=/home/ubuntu/all_websites/target/debug/deps/libsha2-385537d261f23a69.rmeta
> --extern smart_default=/home/ubuntu/all_websites/target/debug/deps/libsmart_default-13b64a7cd8d27eb5.so --extern smol
> _str=/home/ubuntu/all_websites/target/debug/deps/libsmol_str-f10535a685011dcf.rmeta --extern thiserror=/home/ubuntu/a
> ll_websites/target/debug/deps/libthiserror-7a1530efc030e51c.rmeta --extern tokio=/home/ubuntu/all_websites/target/deb
> ug/deps/libtokio-d9d9392c63bfd1e7.rmeta --extern uuid=/home/ubuntu/all_websites/target/debug/deps/libuuid-f2622212b77
> 0dcf3.rmeta --cap-lints allow -C link-arg=-fuse-ld=lld
> ```

The swap file is your best friend in these cases.

**Reply by pikitgb** [reply]
Thank you so much @iganev 

I've decided for now just to use direct calls to Stripe API.
But you suggestions is perfect and in case use useful to others this is what you could use:

```
sudo fallocate -l 4G /swapfile
sudo chmod 600 /swapfile
sudo mkswap /swapfile
sudo swapon /swapfile
sudo swapon --show
echo '/swapfile none swap sw 0 0' | sudo tee -a /etc/fstab
```

**Reply by iganev** [reply]
Yeah, when you absolutely must compile a big codebase on a root vegetable, you'd want to have swap. I found it hard to compile a project using the stripe sdk with less than 16GB memory. 8GB RAM + 8GB swap saves the day, although the pipeline might run forever...

**Reply by stix420** [reply]
I got a similar error when upgrading to 0.38.0, we rolled back to 0.37.1 and were able to compile (0.38.0 was also causing my rust-analyzer+ide to get sigkilled, which ive never seen before)

**Reply by arlyon** [reply]
Please consider trying the next branch if you are struggling with resource usage. While unstable it is the next major version and is getting very close to beta / rc level.

In terms of improvements to compile performance, see this comment: https://github.com/arlyon/async-stripe/discussions/77#discussioncomment-9012903

Beyond that, I would really love some testers and am trying to be responsive to PRs / issues. Alternatively, feel free to reach out directly at arlyon@me.com if you have any feedback on the `next` branch.

**Reply by stix420** [reply]
i didn't realize next was available for testing, i'd be happy to run it in our test enviornments, do you have a rough ETA on when you would consider it production stable?

**Reply by imothee** [reply]
We've been using next in prod for a while now and it's been stable and much cleaner of an API for us, not to mention the build time went from 30mins to 12 and with some smarter caching in docker we now have a 3 min build on a regular GHA.

What more would you like battle tested arlyon/how can we help get a version up on cargo vs. the many many direct git referenced projects in cargo? (which we'd love because we're moving to bazel and the git integration is a bit of overhead vs just using cargo for everything).

**Reply by mzeitlin11** [reply]
Hi @imothee - thanks for trying out the next branch! https://github.com/arlyon/async-stripe/pull/525#issuecomment-2212594059 has a bit of context on current work - from my end would like to easily allow any future necessary breaking changes, but would support the `next` branch being published as an alpha/beta release for users to try out more easily. The main work that would require is updating the `cargo-release` usage (currently just needs to publish a single crate) to publishing the whole workspace. The main questions there would be
- Should all crates be updated in lockstep with a new release, or only as needed if some crates have no changes?
- Are we happy with the new crate names? It would be unfortunate to publish and claim those crate names, then decide different names are better

**Reply by imothee** [reply]
Will continue to think about this but from a first pass i'd likely split the crates into two logical buckets, one that is the core "client" library, includes hand maintained code and all of those would be released and versioned in lockstep even if there was no code change in a sub-path, that just makes the most sense to me. 
Then there's all the generated libs which would recompile and republish whenever an api spec change got bumped (or even versioned to the stripe api changes themselves).
Naming wise, I'd probably try and namespace the generated packages in some way ie. async-stripe-api-* just so it was somewhat clearly differentiated.

**Reply by arlyon** [reply]
I have been sitting on a branch to release an alpha for a while and I think we'll just go for it. Currently versioning everything independently but may revert to lockstep. I'll get a PR up and see if I can get an alpha build on cargo.

**Reply by imothee** [reply]
Amazing, I'm def down to test exhaustively the alpha build, we have 3 large services consuming a large number of stripe endpoints, a bunch of automated tests and a couple eng who can help.

**Reply by imothee** [reply]
We've been using the alpha branch in prod since release and it's been bulletproof, we use almost every package and have thousands of calls to stripe per day and have seen no issues and much faster and cleaner imports and builds! Just wanted to drop the feedback in and say thank you :)

**Comment by arlyon** [comment]
Alpha has been released. Please check it out! You can re-enable serde for specific packages if you want to.
